gensim/models/keyedvectors.py

```diff
@@ -158,264 +163,425 @@
-    double, array, zeros, vstack, sqrt, newaxis, integer, \
-    ndarray, sum as np_sum, prod, argmax
+from numpy import dot, float32 as REAL, \
+    double, array, zeros, vstack, \
+    ndarray, sum as np_sum, prod, argmax, dtype, ascontiguousarray, \
+    frombuffer
 import numpy as np
 
 from gensim import utils, matutils  # utility fnc for pickling, common scipy operations etc
 from gensim.corpora.dictionary import Dictionary
+from gensim.utils import deprecated
 from six import string_types, integer_types
 from six.moves import zip, range
 from scipy import stats
-from gensim.utils import deprecated
-from gensim.models.utils_any2vec import (
-    _save_word2vec_format,
-    _load_word2vec_format,
-    ft_ngram_hashes,
-)
-from gensim.similarities.termsim import TermSimilarityIndex, SparseTermSimilarityMatrix
 
-#
-# For backwards compatibility, see https://github.com/RaRe-Technologies/gensim/issues/2201
-#
-from gensim.models.deprecated.keyedvectors import EuclideanKeyedVectors  # noqa
 
 logger = logging.getLogger(__name__)
 
 
-class Vocab(object):
-    """A single vocabulary item, used internally for collecting per-word frequency/sampling info,
-    and for constructing binary trees (incl. both word leaves and inner nodes).
+KEY_TYPES = (string_types, integer_types, np.integer)
 
-    """
-    def __init__(self, **kwargs):
-        self.count = 0
-        self.__dict__.update(kwargs)
 
-    def __lt__(self, other):  # used for sorting in a priority queue
-        return self.count < other.count
-
-    def __str__(self):
-        vals = ['%s:%r' % (key, self.__dict__[key]) for key in sorted(self.__dict__) if not key.startswith('_')]
-        return "%s(%s)" % (self.__class__.__name__, ', '.join(vals))
+class KeyedVectors(utils.SaveLoad):
+    def __init__(self, vector_size, mapfile_path=None):
+        """Mapping between keys (such as words)  and vectors for :class:`~gensim.models.Word2Vec`
+        and related models.
 
+        Used to perform operations on the vectors such as vector lookup, distance, similarity etc.
 
-class BaseKeyedVectors(utils.SaveLoad):
-    """Abstract base class / interface for various types of word vectors."""
-    def __init__(self, vector_size):
-        self.vectors = zeros((0, vector_size), dtype=REAL)
-        self.vocab = {}
+        """
         self.vector_size = vector_size
-        self.index2entity = []
+        self.vectors = zeros((0, vector_size), dtype=REAL)  # fka (formerly known as) syn0
+        self.norms = None
 
-    def save(self, fname_or_handle, **kwargs):
-        super(BaseKeyedVectors, self).save(fname_or_handle, **kwargs)
+        self.index_to_key = []  # fka index2entity or index2word
+        self.key_to_index = {}
 
-    @classmethod
-    def load(cls, fname_or_handle, **kwargs):
-        return super(BaseKeyedVectors, cls).load(fname_or_handle, **kwargs)
+        self.expandos = {}  # dynamically-expandable per-vector named, numpy-typed attributes
```

Comment:
"Dynamically-expandable" and "numpy-typed" seems like a contradiction.

Check: What is this actually for, why is it here?

Reply:
Discussed further at: https://github.com/RaRe-Technologies/gensim/pull/2698#issuecomment-654530822